### PR TITLE
Adding performance file deletion test

### PR DESCRIPTION
**Pull request by tonynguien**
### Motivation and Context

File deletion on object storage can take a long time. Adding a test so we can track and compare file deletion between block devices and object storage.

### Description

Adding a new test that do the following:
- Creating a single 600GB file
- 8K randomly writing 5GB worth of data to file
- Time the delete operation, i.e. rm(1)

### How Has This Been Tested?

Block devices run, file deletion took ~3.6 seconds
```Test: /usr/share/zfs/zfs-tests/tests/perf/regression/file_deletion (run as root) [11:30] [PASS]
00:24:18.55 removed all registered handlers
00:24:18.55 SUCCESS: zinject -c all
00:24:18.60 SUCCESS: zpool destroy -f perfpool
00:24:18.60 NOTE: zpool create -f perfpool nvme4n1 nvme5n1 nvme6n1
00:24:19.42 SUCCESS: zfs create -o recsize=8k -o compress=on -o checksum=edonr -o relatime=on perfpool/fs1
00:24:19.43 NOTE: Random writes
00:24:19.73 job: (g=0): rw=randwrite, bs=(R) 8192B-8192B, (W) 8192B-8192B, (T) 8192B-8192B, ioengine=psync, iodepth=1
00:24:19.73 fio-3.1
00:24:19.73 Starting 1 thread
00:24:19.73 job: Laying out IO file (1 file / 575999MiB)
00:35:43.51
00:35:43.51 job: (groupid=0, jobs=1): err= 0: pid=5430: Fri Sep 17 00:35:43 2021
00:35:43.51   write: IOPS=958, BW=7669KiB/s (7853kB/s)(5120MiB/683644msec)
00:35:43.51     clat (usec): min=559, max=73331, avg=1040.30, stdev=705.01
00:35:43.51      lat (usec): min=559, max=73331, avg=1040.45, stdev=705.02
....
00:35:43.51 Run status group 0 (all jobs):
00:35:43.51   WRITE: bw=7669KiB/s (7853kB/s), 7669KiB/s-7669KiB/s (7853kB/s-7853kB/s), io=5120MiB (5369MB), run=683644-683644msec
00:35:43.54 SUCCESS: fio /usr/share/zfs/zfs-tests/tests/perf/fio/random_writes_fill.fio
00:35:43.55 NOTE: Collecting backend IO stats with lun list :
00:35:43.55 NOTE: Removing file
00:35:43.56 NOTE: DIRECTORY:  /perfpool/fs1
00:35:44.74 SUCCESS: zinject -a
00:35:48.35 SUCCESS: rm /perfpool/fs1/file0
00:35:48.35 NOTE: /perfpool/fs1/file0 deletion took: 3.61391610336303709 secs
```

Block object run, delete took ~24 seconds.
```
Test: /usr/share/zfs/zfs-tests/tests/perf/regression/setup (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/perf/regression/file_deletion (run as root) [01:23] [PASS]
00:46:15.69 removed all registered handlers
00:46:15.69 SUCCESS: zinject -c all
00:46:15.89 SUCCESS: zpool destroy -f perfpool
00:46:15.89 NOTE: zpool create -o object-endpoint=https://s3-us-west-2.amazonaws.com -o object-region=us-west-2 -o object-credentials-profile=default perfpool s3 cloudburst-data-2
00:46:23.44 SUCCESS: zfs create -o recsize=8k -o compress=on -o checksum=edonr -o relatime=on perfpool/fs1
00:46:23.45 NOTE: Random writes
00:46:23.74 job: (g=0): rw=randwrite, bs=(R) 8192B-8192B, (W) 8192B-8192B, (T) 8192B-8192B, ioengine=psync, iodepth=1
00:46:23.74 fio-3.1
00:46:23.74 Starting 1 thread
00:46:23.74 job: Laying out IO file (1 file / 614400MiB)
00:47:04.46
00:47:04.46 job: (groupid=0, jobs=1): err= 0: pid=6176: Fri Sep 17 00:47:04 2021
00:47:04.46   write: IOPS=16.1k, BW=126MiB/s (132MB/s)(5120MiB/40589msec)
00:47:04.46     clat (usec): min=12, max=17006, avg=59.26, stdev=127.06
00:47:04.46      lat (usec): min=12, max=17006, avg=59.39, stdev=127.12
....
00:47:04.46 Run status group 0 (all jobs):
00:47:04.46   WRITE: bw=126MiB/s (132MB/s), 126MiB/s-126MiB/s (132MB/s-132MB/s), io=5120MiB (5369MB), run=40589-40589msec
00:47:04.51 SUCCESS: fio /usr/share/zfs/zfs-tests/tests/perf/fio/random_writes_fill.fio
00:47:04.53 NOTE: Collecting backend IO stats with lun list :
00:47:04.57 NOTE: Removing file
00:47:04.57 NOTE: DIRECTORY:  /perfpool/fs1
00:47:06.15 SUCCESS: zinject -a
00:47:30.00 SUCCESS: rm /perfpool/fs1/file0
00:47:30.00 NOTE: /perfpool/fs1/file0 deletion took: 23.8508140926361084 secs
```

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: tests/zfs-tests/tests/perf/regression/file_deletion.ksh [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+# Variables for use by fio.
+export PERF_RUNTIME=${PERF_RUNTIME:-$PERF_RUNTIME_NIGHTLY}
+export PERF_RANDSEED=${PERF_RANDSEED:-'1234'}
+export PERF_COMPPERCENT=${PERF_COMPPERCENT:-'66'}
+export PERF_COMPCHUNK=${PERF_COMPCHUNK:-'4096'}
+export PERF_RUNTYPE=${PERF_RUNTYPE:-'nightly'}
+export PERF_NTHREADS=1
+export PERF_NTHREADS_PER_FS=${PERF_NTHREADS_PER_FS:-'0'}
+export PERF_SYNC_TYPES=${PERF_SYNC_TYPES:-'1'}
+export PERF_IOSIZES=${PERF_IOSIZES:-'8k'}
```
Comment (jwk404) [inline tests/zfs-tests/tests/perf/regression/file_deletion.ksh:59]: Some of these are already set in perflib.shlib from a recent change. Where they're identical, we can omit them here.

File: tests/zfs-tests/tests/perf/regression/file_deletion.ksh [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	    "$perf_record_cmd" "perf_writes"
+	)
+else
+	export collect_scripts=(
+	    "$PERF_SCRIPTS/io.d $PERFPOOL $lun_list 1" "io"
+	    "vmstat -T d 1" "vmstat"
+	    "mpstat -T d 1" "mpstat"
+	    "iostat -T d -xcnz 1" "iostat"
+	)
+fi
```
Comment (jwk404) [inline tests/zfs-tests/tests/perf/regression/file_deletion.ksh:76]: I think you can delete this, as it's duplicated below.

File: tests/zfs-tests/tests/perf/regression/file_deletion.ksh [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+export PERF_RUNTIME=10
+if is_linux; then
+	typeset perf_record_cmd="perf record --call-graph dwarf,8192 -F 49 -agq \
+	    -o /dev/stdout -- sleep ${PERF_RUNTIME}"
+
+	export collect_scripts=(
+	    "zpool iostat -lpvyL $PERFPOOL 1" "zpool.iostat"
+	    "iostat -tdxyz 1" "iostat"
+	    "arcstat 1" "arcstat"
+	    "dstat -at --nocolor 1" "dstat"
```
Comment (jwk404) [inline tests/zfs-tests/tests/perf/regression/file_deletion.ksh:94]: dstat can produce csv output. Should we do that here for easier parsing later?
(I'm assuming the csv goes to stdout here, but not 100% sure about that)

**Review by jwk404 — CHANGES_REQUESTED** [review]

File: tests/zfs-tests/tests/perf/regression/file_deletion.ksh (hunk 2 above)
Comment (tonynguien) [reply tests/zfs-tests/tests/perf/regression/file_deletion.ksh:76]: I deleted the below section as we want to collect data during the random writing and the deletion.

File: tests/zfs-tests/tests/perf/regression/file_deletion.ksh (hunk 1 above)
Comment (tonynguien) [reply tests/zfs-tests/tests/perf/regression/file_deletion.ksh:59]: We don't yet have that branch in the object storage branch.

**Review by jwk404 — APPROVED** [review]

File: tests/zfs-tests/tests/perf/regression/file_deletion.ksh (hunk 3 above)
Comment (tonynguien) [reply tests/zfs-tests/tests/perf/regression/file_deletion.ksh:94]: Yeah, I think we definitely want use csv when we're ready to working on parsing. Until then, I think it's nice to look at content for dstat same as other perf stat files.

File: tests/zfs-tests/tests/perf/regression/file_deletion.ksh [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	export collect_scripts=(
+	    "$PERF_SCRIPTS/io.d $PERFPOOL $lun_list 1" "io"
+	    "vmstat -T d 1" "vmstat"
+	    "mpstat -T d 1" "mpstat"
+	    "iostat -T d -xcnz 1" "iostat"
+	)
+fi
+do_collect_scripts delete
+
+log_note "Removing file"
```
Comment (grwilson) [inline tests/zfs-tests/tests/perf/regression/file_deletion.ksh:100]: Prior to starting the remove, you may need to wait for the I/Os to complete. I think `zpool sync <poolname>` will do that for you.

Comment (tonynguien) [reply tests/zfs-tests/tests/perf/regression/file_deletion.ksh:100]: After chatting with George, we'll go with:
1) async write by default (it's currently defaulting to sync)
2) zpool sync prior to rm
This way we can speed up the writes.

Comment (tonynguien) [reply tests/zfs-tests/tests/perf/regression/file_deletion.ksh:102]: Using async, the the 5GB writes now takes ~2 minutes compared to >10 minutes with prior sync.

**Review by grwilson — APPROVED** [review]

**Pull request merged**